clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,285 @@
+
+	pos, err := positions.Get(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "get positions")
+	}
+
+	var decoder *encoding.Decoder
+	if encodingFormat != "" {
+		level.Info(logger).Log("msg", "decompresser will decode messages", "from", encodingFormat, "to", "UTF8")
+		encoder, err := ianaindex.IANA.Encoding(encodingFormat)
+		if err != nil {
+			return nil, errors.Wrap(err, "error doing IANA encoding")
+		}
+		decoder = encoder.NewDecoder()
+	}
+
+	decompresser := &decompresser{
+		metrics:   metrics,
+		logger:    logger,
+		handler:   api.AddLabelsMiddleware(model.LabelSet{FilenameLabel: model.LabelValue(path)}).Wrap(handler),
+		positions: positions,
+		path:      path,
+		running:   atomic.NewBool(false),
+		posquit:   make(chan struct{}),
+		posdone:   make(chan struct{}),
+		done:      make(chan struct{}),
+		position:  pos,
+		decoder:   decoder,
+	}
+
+	go decompresser.readLines()
+	go decompresser.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompresser, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompresser.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
+	ext := filepath.Ext(f.Name())
+	var decompressLib string
+
+	if strings.Contains(ext, "gz") { // .gz, .tar.gz
+		decompressLib = "compress/gzip"
+		reader, err = gzip.NewReader(f)
+	} else if ext == ".z" {
+		decompressLib = "compress/zlib"
+		reader, err = zlib.NewReader(f)
+	} else if ext == ".zip" {
+		decompressLib = "compress/flate"
+		reader = flate.NewReader(f)
+	} else if ext == ".bz2" {
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Info(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("file %q with unsupported extension", f.Name())
+}
+
+func (t *decompresser) updatePosition() {
+	positionSyncPeriod := t.positions.SyncPeriod()
+	positionWait := time.NewTicker(positionSyncPeriod)
+	defer func() {
+		positionWait.Stop()
+		level.Info(t.logger).Log("msg", "position timer: exited", "path", t.path)
+		close(t.posdone)
+	}()
+
+	for {
+		select {
+		case <-positionWait.C:
+			if err := t.MarkPositionAndSize(); err != nil {
+				level.Error(t.logger).Log("msg", "position timer: error getting position and/or size, stopping decompresser", "path", t.path, "error", err)
+				return
+			}
+		case <-t.posquit:
+			return
+		}
+	}
+}
+
+// readLines read all existing lines of the given compressed file.
+//
+// It first decompress the file as a whole using a reader and then it will iterate
+// over its chunks, separated by '\n'.
+// During each iteration, the parsed and decoded log line is then sent to the API with the current timestamp.
```

Comment:
This is what I was kinda worried about seeing...
If the file is compressed, it could potentially store a huge amount of data which we need to entirely load into memory to iterate over.

Is there a more resource-efficient way we could do this?

Reply:
Totally @dannykopping  ... compression ratios >99% are common when thinking about uniformly structured logs like access or event logs. Reading a whole file into memory would totally not be an option. When suggesting implementing this along the lines of piping data via `zcat` into an existing tool (https://github.com/grafana/loki/issues/5956#issuecomment-1185366990). And isn't that exactly what the mountReader method does? It gets the right type of reader (which is able to read a compressed data stream of various algorithms) and then hands that off to be read from line by line. But there is not slurping into memory, is there?

Reply:
We tested this together, and found that it is indeed allocating a lot of memory **but** the GC acts as it should, and the memory doesn't grow unbounded. Dylan also verified that the files are decompressed progressively, so at no time do we try to load the whole file into memory at once, so the overall memory usage will not spike very high (we use small buffers when reading the compressed files). We should add a note in the docs to specify that GC frequency will increase dramatically when decompressing large files, and this will cause CPU usage to rise.